File: src/paths.ts
import { homedir } from "node:os";
import { join } from "node:path";

export const getHomeDir = (): string => join(homedir(), ".railgun");

export const pathsForHome = (home: string) => ({
  config: join(home, "config.json"),
  token: join(home, "devin-token"),
  state: join(home, "state.db"),
  soul: join(home, "SOUL.md"),
  trust: join(home, "trust.json"),
  extensions: join(home, "extensions"),
  cron: join(home, "cron", "jobs.json"),
  cronLogs: join(home, "cron", "logs"),
  skills: join(home, "skills"),
} as const);

const paths = pathsForHome(getHomeDir());

export const CONFIG_PATH = paths.config;
export const TOKEN_PATH = paths.token;
export const STATE_PATH = paths.state;
export const SOUL_PATH = paths.soul;
export const TRUST_PATH = paths.trust;
export const EXTENSIONS_PATH = paths.extensions;
export const CRON_PATH = paths.cron;
export const SKILLS_PATH = paths.skills;
export const CRON_LOGS_PATH = paths.cronLogs;

File: crates/railgun-backend/src/paths.rs
use std::path::PathBuf;

#[derive(Clone, Debug)]
pub struct RailgunPaths {
    pub home: PathBuf,
    pub config: PathBuf,
    pub token: PathBuf,
    pub state: PathBuf,
    pub soul: PathBuf,
    pub extensions: PathBuf,
    pub cron: PathBuf,
    pub skills: PathBuf,
}

impl RailgunPaths {
    pub fn discover() -> anyhow::Result<Self> {
        let user_home = std::env::var_os("HOME")
            .map(PathBuf::from)
            .ok_or_else(|| anyhow::anyhow!("HOME is not set"))?;
        Ok(Self::for_user_home(user_home))
    }

    pub fn for_user_home(user_home: impl Into<PathBuf>) -> Self {
        let home = user_home.into().join(".railgun");
        Self {
            config: home.join("config.json"),
            token: home.join("devin-token"),
            state: home.join("state.db"),
            soul: home.join("SOUL.md"),
            extensions: home.join("extensions"),
            cron: home.join("cron/jobs.json"),
            skills: home.join("skills"),
            home,
        }
    }
}
